### DISCUSSION TITLE
youtube feeds parsing problems

### DISCUSSION THREAD

**Discussion by nsklaus** [ViennaRSS/vienna-rss#1431] [Ideas]
other rss clients are able to display the video thumbnail alongside the video description. vienna only show the video description text but fails to show the thumbnail/video preview/still frame. instead it says there is an enclosed file, but i think it is an error of parsing. it seems in the case of youtube subscriptions, it download the whole html page and make it an attachment to the feed.

also i noticed the feed date is getting errors. i'm getting old feed from september 2020 to popup as new in december 2020. here are the steps i did: i deleted my old subscriptions and started fresh today, january the 2nd of 2021. i clicked "refresh subscriptions" button to make the initial sync, and i saw some old videos from september being marked as new from december. and not only on one subscription only but many of them showed this problem. vienna is incorrectly pulling the feed date.

as an example of these problem please see the screenshot, 
vienna says the highlighted feed is from 29/12/2020
the link to this video is: https://www.youtube.com/watch?v=a8-f7y2Jkr4
when i go to that page, i can see the video is in fact from 06/11/2020
![Screenshot 2021-01-02 at 04 34 15](https://user-images.githubusercontent.com/6810861/103450281-1a0fcf00-4cb4-11eb-9d23-b5a1701901a6.jpg)

also finally, it would be nice if vienna could send links to some external player, like vlc or mpv or iina.

test environment: vienna version 3.7.0 :5698f89d: (7429) on macos bigsur.

**Comment by Eitot** [comment]
This is from the RSS feed:
```xml
 <entry>
  <id>yt:video:a8-f7y2Jkr4</id>
  <yt:videoId>a8-f7y2Jkr4</yt:videoId>
  <yt:channelId>UC5rT7F0PGNuD54rJ9kzgWzw</yt:channelId>
  <title>Off Grid Cider Making 2020</title>
  <link rel="alternate" href="https://www.youtube.com/watch?v=a8-f7y2Jkr4"/>
  <author>
   <name>Kris Harbour Natural Building</name>
   <uri>https://www.youtube.com/channel/UC5rT7F0PGNuD54rJ9kzgWzw</uri>
  </author>
  <published>2020-11-06T14:11:10+00:00</published>
  <updated>2020-12-29T17:08:45+00:00</updated>
  <media:group>
   <media:title>Off Grid Cider Making 2020</media:title>
   <media:content url="https://www.youtube.com/v/a8-f7y2Jkr4?version=3" type="application/x-shockwave-flash" width="640" height="390"/>
   <media:thumbnail url="https://i2.ytimg.com/vi/a8-f7y2Jkr4/hqdefault.jpg" width="480" height="360"/>
   <media:description>Its the time of year again, Cider making. It was a good year for apples. we made 35L of cider and 35L of apple juice for the freezer.</media:description>
   <media:community>
    <media:starRating count="3659" average="4.96" min="1" max="5"/>
    <media:statistics views="57156"/>
   </media:community>
  </media:group>
 </entry>
```

The parser is working correctly. The "Date" column shows the date upon which the content was last updated, which in this case is 29 December 2020. It seems that YouTube refers to the Flash version of the video as the feed entry's "enclosure". This is what Vienna uses to have compatibility with the RSS `<enclosure>` element.

To support video thumbnails, Vienna would have to add support for the `<media:thumbnail>` element and make the table view capable of showing larger images.

**Comment by nsklaus** [comment]
the reasons you put forward are logic and all, but i just wanted to mention that in real life cases it leads to degraded user experience and that in comparison, other newsfeed readers don't seems to exhibit the same behaviour.
for example, "reeder", "leaf" and "news explorer" don't show these dates and thumbnails problems. furthermore, vienna seems capable enough to show thumbnails in other kind of feeds beside youtube ones. so considering this all, i thought it might be worth reporting.

maybe adding an exception case for youtube feeds so it would take the "published" date instead of the "updated" one ?

also you seem to indicate vienna doesn't support showing larger image ? looking at this feed here:
![Screenshot 2021-01-02 at 15 46 38](https://user-images.githubusercontent.com/6810861/103459801-dea3ed80-4d11-11eb-911f-e747837f29c6.jpg)

it seems bigger to me than most of the youtube videos "thumbnail/preview/still frame" usualy are. for reference purpose, here's what it look like on reeder:

![Screenshot 2021-01-02 at 15 57 32](https://user-images.githubusercontent.com/6810861/103460031-69d1b300-4d13-11eb-8f83-f643d8385230.jpg)

**Comment by Eitot** [comment]
This is not the same. This is how the feed of Macrumors looks:
```xml
<item>
<title>Apple Removes 39,000 Games From China App Store</title>
<link>https://www.macrumors.com/2020/12/31/apple-removes-39000-game-apps-from-china-store/</link>
<comments>https://forums.macrumors.com/threads/apple-removes-39-000-games-from-china-app-store.2278090</comments>
<pubDate>Thu, 31 Dec 2020 04:24:10 PST</pubDate>
<guid isPermaLink="false">https://www.macrumors.com/2020/12/31/apple-removes-39000-game-apps-from-china-store/</guid>
<dc:creator>Tim Hardwick</dc:creator>
<category>App Store</category>
<category>China</category>
<description>Apple on Thursday removed nearly 39,000 apps from its Chinese App Store due to the apps lacking an official license from local regulators, reports &lt;em&gt;&lt;a href="https://www.reuters.com/article/us-apple-china-games-idUSKBN2950P1"&gt;Reuters&lt;/a&gt;&lt;/em&gt;.
&lt;br/&gt;

&lt;br/&gt;
&lt;img src="https://images.macrumors.com/article-new/2020/03/appstore.jpg" alt="" width="2000" height="1032" class="aligncenter size-full wp-image-730995" /&gt;
&lt;br/&gt;
...
</description>
</item>
```

You can see here that the large image is directly embedded in the `<description>` of the feed item. Vienna's article viewer renders this `<description>` as HTML, including all embedded images.

In the case of YouTube, the thumbnail image is instead declared as an enclosure (i.e. attachment) to the feed entry, using a non-standard `<media:thumbnail>` element. To add support for this, Vienna would have to:
1. Extend the parser to parse `<media:thumbnail>`
2. Extend the data model to have a field for the thumbnail
3. Extend the view model to allow for thumbnail images to be shown in the article list

I have marked this as a feature request.

**Comment by nsklaus** [comment]
ok i understand, thank you for the explanations and for making this a feature request.
while waiting for this to be correctly implemented, i went ahead and played a bit with the code, and made these for myself the quick and ugly way i guess.

this is probably one of the worst piece of code i have done.
i had never touched objective_c before.
but, fiddling with `RichXMLParser.m`, 
around block starting at line 761, [here](https://github.com/ViennaRSS/vienna-rss/blob/1aab78f10320350ae955b5b2fa14c7766755f28e/Vienna/Sources/Parsing/RichXMLParser.m#L760)

i was able to add support for showing the youtube feed preview image and correcting the date too. so far i see no bad side effects on the other feeds.

![Screenshot 2021-01-03 at 01 46 00](https://user-images.githubusercontent.com/6810861/103469464-8b5b8a80-4d65-11eb-9fd7-92444a7c89ba.jpg)

**Comment by nsklaus** [comment]
i don't think anyone would want to use my "code" as it's probably really crappy quality. i suppose providing pull request for this wouldn't be acceptable, quality wise, as it's probably done in the worst way possible but,
just in case it can serve for reference purpose:

![Screenshot 2021-01-02 at 22 41 34](https://user-images.githubusercontent.com/6810861/103467112-f730f980-4d4b-11eb-84eb-2c8ef005cf27.jpg)
